src/type-layout.md

```diff
@@ -0,0 +1,241 @@
+# Type Layout
+
+The layout of a type is the way the size, alignment, and the offsets of any
+fields and discriminants for the values of that type.
+
+While specific releases of the compiler will have the same layout for types,
+there is a lot of room for new versions of the compiler to do different things.
+Instead of trying to document exactly what is done, we only document what is
+guaranteed today.
+
+## Size and Alignment
+
+All values have an alignment and size.
+
+The *alignment* of a value specifies what addresses are valid to store the value
+at. A value of alignment `n` must only be stored at an address that is a
+multiple of n. For example, a value with an alignment of 2 must be stored at an
+even address, while a value with an alignment of 1 can be stored at any address.
+Alignment is measured in bytes, and must be at least 1, and always a power of 2.
+The alignment of a value can be checked with the [`align_of_val`] function.
+
+The *size* of a value is the offset in bytes between successive elements in an
+array with that item type including alignment padding. The size of a value is
+always a multiple of its alignment. The size of a value can be checked with the
+[`size_of_val`] function.
+
+Types where all values have the same size and alignment known at compile time
+implement the [`Sized`] trait and can be checked with the [`size_of`] and
+[`align_of`] functions. Types that are not [`Sized`] are known as [dynamically
+sized types]. Since all values of a `Sized` type share the same size and
+alignment, we refer to those shared values as the size of the type and the
+alignment of the type respectively.
+
+## Primitive Data Layout
+
+The size of most primitives is given in this table.
+
+Type | `size_of::\<Type>()`
+- | - | -
+bool | 1
+u8 | 1
+u16 | 2
+u32 | 4
+u64 | 8
+i8 | 1
+i16 | 2
+i32 | 4 
+i64 | 8
+f32 | 4
+f64 | 8
+char | 4
+
+`usize` and `isize` have a size big enough to contain every address on the
+target platform. For example, on a 32 bit target, this is 4 bytes and on a 64
+bit target, this is 8 bytes.
+
+Most primitives are generally aligned to their size, although this is
+platform-specific behavior. In particular, on x86 u64 and f64 may be only
+aligned to 32 bits.
+
+## Pointers and References Layout
+
+Pointers and references have the same layout. Mutability of the pointer or
+reference does not change the layout.
+
+Pointers to sized types have the same size and alignment as `usize`.
+
+Pointers to unsized types are sized. The size and alignemnt is guaranteed to be
+at least equal to the size and alignment of a pointer.
+
+> Note: Though you should not rely on this, all pointers to <abbr
+> title="Dynamically Sized Types">DSTs</abbr> are currently twice the size of
+> the size of `usize` and have the same alignment.
+
+## Array Layout
+
+Arrays are laid out so that the `nth` element of the array is offset from the
+start of the array by `n * the size of the type` bytes. An array of `[T; n]`
```

Comment:
Could do. Will consider it.

Reply:
I tried, and I realized that I'd need a second variant (other than `n`) for the size of the array since the `n` in the first sentence is distinct from the `n` in the second. And I don't know how to introduce it easily. "nth" is almost its own word at this point, so showing "ith" or "xth" would be confusing.